wheel/python/clvm_rs/casts.py

```diff
@@ -0,0 +1,170 @@
+"""
+Some utilities to cast python types to and from clvm.
+"""
+
+from typing import Callable, List, SupportsBytes, Tuple, Union, cast
+
+from .clvm_storage import CLVMStorage, is_clvm_storage
+
+
+AtomCastableType = Union[
+    bytes,
+    str,
+    int,
+    SupportsBytes,
+]
+
+
+CastableType = Union[
+    AtomCastableType,
+    List["CastableType"],
+    Tuple["CastableType", "CastableType"],
+    CLVMStorage,
+]
+
+
+NULL_BLOB = b""
+
+
+def generate_working_memview_or_bytes():
+    # in python3.7 you can't go 'isinstance(b"", SupportsBytes)'
+
+    # this one works in py38+
+    def memview_or_bytes_py38_or_later(o):
+        return isinstance(o, (memoryview, SupportsBytes))
+
+    # this one works in py37
+    def memview_or_bytes_py37(o):
+        return getattr(o, "__bytes__", None) is not None
+
+    try:
+        memview_or_bytes_py38_or_later(b"")
+        return memview_or_bytes_py38_or_later
+    except TypeError:
+        pass
+    return memview_or_bytes_py37
+
+
+memview_or_bytes = generate_working_memview_or_bytes()
+
+
+def int_from_bytes(blob):
+    """
+    Convert a bytes blob encoded as a clvm int to a python int.
+    """
+    size = len(blob)
+    if size == 0:
+        return 0
+    return int.from_bytes(blob, "big", signed=True)
+
+
+def int_to_bytes(v) -> bytes:
+    """
+    Convert a python int to a blob that encodes as this integer in clvm.
+    """
+    byte_count = (v.bit_length() + 8) >> 3
+    if v == 0:
+        return b""
+    r = v.to_bytes(byte_count, "big", signed=True)
+    # make sure the string returned is minimal
+    # ie. no leading 00 or ff bytes that are unnecessary
+    while len(r) > 1 and r[0] == (0xFF if r[1] & 0x80 else 0):
+        r = r[1:]
+    return r
+
+
+def to_atom_type(v: AtomCastableType) -> bytes:
+    """
+    Convert an `AtomCastableType` to `bytes`. This for use with the
+    convenience function `Program.to`.
+    """
+    if isinstance(v, bytes):
+        return v
+    if isinstance(v, str):
+        return v.encode()
+    if isinstance(v, int):
+        return int_to_bytes(v)
+    if memview_or_bytes(v):
+        return bytes(v)
+
+    raise ValueError("can't cast %s (%s) to bytes" % (type(v), v))
+
+
+def to_clvm_object(
```

Comment:
I think you should have a unit test for this function. It should also have a return type annotation.
I imagine this is a dangerous call to make on untrusted input. It might make sense to document that.

Reply:
The only dangerous case I can think of is a recursive list structure (as in `t = []; t.append(t); to_clvm_object(t)`). Any other cases?

I don't see this being used on untrusted input much, since untrusted input would generally be in clvm form already, rather than python objects.

Reply:
Unit tests are in https://github.com/Chia-Network/clvm_rs/blob/0ddb2668434de8a19222eed56fef060f9e7ca44b/wheel/python/tests/test_as_python.py (I probably should rename this file).

Reply:
I used to assume that the wallet would only manipulate clvm structures and programs that it had produced itself, and therefore was safe. I don't think that's the case anymore and I think it's risky to assume that code that is only called on trusted structures *today* will not be used on untrusted structures in the future. I believe there are plenty of examples of this happening in the wallet.

Reply:
The object that comes back from `.run` conforms to `CLVMStorage`. This function should notice and return the original object with no changes.

Reply:
> I just realized this is converting in the opposite direction though, so probably lower risk. As long as people don't get a lazily convertible output from a malicious program, and then convert it back to clvm via this function.

I removed `.as_python` from the API for just this reason, and now point people to safer alternatives like `.as_iter`, `.at` and `.at_many`.